pkg/segment/aggregations/segaggs.go

```diff
@@ -196,26 +206,247 @@ func performAggOnResult(nodeResult *structs.NodeResult, agg *structs.QueryAggreg
+							newResults = append(newResults, bucketResult)
+						} else if headExpr.Keeplast {
+							newResults = append(newResults, bucketResult)
+							headExpr.Done = true
+							break
+						} else {
+							headExpr.Done = true
+							break
+						}
+					} else {
+						return fmt.Errorf("performConditionalHeadOnHistogram: Error while evaluating expression on histogram, err: %v", err)
+					}
+				} else {
+					if result {
+						newResults = append(newResults, bucketResult)
+					} else {
+						// false condition so adding last record if keeplast
+						if headExpr.Keeplast {
+							newResults = append(newResults, bucketResult)
+						}
+						headExpr.Done = true
+						break
+					}
+				}
+
+				if headExpr.MaxRows > 0 && headExpr.RowsAdded == headExpr.MaxRows {
+					headExpr.Done = true
+					break
+				}
+
+			}
+		}
+
+		aggregationResult.Results = newResults
+	}
+
+	return nil
+}
+
+func addRecordToHeadExpr(headExpr *structs.HeadExpr, record map[string]interface{}, recordKey string, hasSort bool) {
+	headExpr.RowsAdded++
+	if hasSort {
+		// we do not need to accumulate the results in case of sort
+		return
+	}
+	headExpr.ResultRecords = append(headExpr.ResultRecords, record)
+	headExpr.ResultRecordKeys = append(headExpr.ResultRecordKeys, recordKey)
+	delete(headExpr.SegmentRecords, recordKey)
+}
+
+func processSegmentRecordsForHeadExpr(headExpr *structs.HeadExpr, recordMap map[string]map[string]interface{}, recordIndexInFinal map[string]int, hasSort bool) error {
+	fieldsInExpr := headExpr.BoolExpr.GetFields()
+	currentOrder := make([]string, len(recordMap))
+
+	for recordKey := range recordMap {
+		idx, exist := recordIndexInFinal[recordKey]
+		if !exist {
+			return fmt.Errorf("processSegmentRecordsForHeadExpr: Index not found in recordIndexInFinal for record: %v", recordKey)
+		}
+		currentOrder[idx] = recordKey
+	}
+
+	for _, recordKey := range currentOrder {
+		rec, exist := recordMap[recordKey]
+		if !exist {
+			return fmt.Errorf("processSegmentRecordsForHeadExpr: record %v not found in segment records", recordKey)
+		}
+
+		fieldToValue := make(map[string]segutils.CValueEnclosure, 0)
+		err := getRecordFieldValues(fieldToValue, fieldsInExpr, rec)
+		if err != nil {
+			return fmt.Errorf("processSegmentRecordsForHeadExpr: Error while retrieving values, err: %v", err)
+		}
+
+		result, err := headExpr.BoolExpr.Evaluate(fieldToValue)
+		if err != nil {
+			nullFields, err := headExpr.BoolExpr.GetNullFields(fieldToValue)
+			if err == nil && len(nullFields) > 0 {
+				// evaluation failed due to null fields
+				if headExpr.Null {
+					addRecordToHeadExpr(headExpr, rec, recordKey, hasSort)
+				} else if headExpr.Keeplast {
+					addRecordToHeadExpr(headExpr, rec, recordKey, hasSort)
+					headExpr.Done = true
+					break
+				} else {
+					headExpr.Done = true
+					break
+				}
+			} else {
+				return fmt.Errorf("processSegmentRecordsForHeadExpr: Error while evaluating expression, err: %v", err)
```

Comment:
On line 308 you make a new `err` variable, so on line 322 you're printing that new `err` instead of the `err` used on 307. I don't think that's what you intended

Reply:
addressed in the latest commit